src/target/riscv/batch.c

```diff
@@ -340,16 +350,20 @@ void riscv_batch_add_dmi_write(struct riscv_batch *batch, uint64_t address, uint
 	batch->used_scans++;
 }
 
-size_t riscv_batch_add_dmi_read(struct riscv_batch *batch, uint64_t address,
+size_t riscv_batch_add_dmi_read(struct riscv_batch *batch, uint32_t address,
 		enum riscv_scan_delay_class delay_class)
 {
+	// TODO: Check that the bit width of "address" is no more than dtmcs.abits,
+	// otherwise return an error (during batch creation or when the batch is executed).
+
 	assert(batch->used_scans < batch->allocated_scans);
 	struct scan_field *field = batch->fields + batch->used_scans;
-	field->num_bits = riscv_get_dmi_scan_length(batch->target);
+	field->num_bits = get_dmi_scan_length(batch->target);
+	assert(field->num_bits <= DMI_SCAN_MAX_BIT_LENGTH);
 	field->out_value = (void *)(batch->data_out + batch->used_scans * DMI_SCAN_BUF_SIZE);
 	field->in_value  = (void *)(batch->data_in  + batch->used_scans * DMI_SCAN_BUF_SIZE);
```

Comment:
These casts also seem redundant:
```suggestion
	field->out_value = batch->data_out + batch->used_scans * DMI_SCAN_BUF_SIZE;
	field->in_value  = batch->data_in  + batch->used_scans * DMI_SCAN_BUF_SIZE;
```

Reply:
Addressed: `(void *)` casts removed.